reconcile/utils/slo_document_manager.py

```diff
@@ -0,0 +1,204 @@
+                    namespace.prometheus_access.username
+                )
+                password = self.secret_reader.read_secret(
+                    namespace.prometheus_access.password
+                )
+
+                prom_url = namespace.prometheus_access.url
+        else:
+            prom_url = namespace.namespace.cluster.prometheus_url
+            if not namespace.namespace.cluster.automation_token:
+                raise Exception(
+                    f"cluster {namespace.namespace.cluster.name} does not have automation token set"
+                )
+            prom_token = self.secret_reader.read_secret(
+                namespace.namespace.cluster.automation_token
+            )
+        return PromCredentials(
+            prom_url=prom_url,
+            prom_token=prom_token,
+            is_basic_auth=is_basic_auth,
+            username=username,
+            password=password,
+        )
+
+    def _extract_current_slo_value(self, data: Any, slo_name: str) -> float:
+        result = data["data"]["result"]
+        if not result:
+            raise EmptySLOResult(
+                f"prometheus returned empty result for SLO: {slo_name} of SLO document: {self.slo_document_name}"
+            )
+        slo_value = result[0]["value"]
+        if not slo_value:
+            raise EmptySLOValue(
+                f"prometheus returned empty SLO value for SLO: {slo_name} of SLO document: {self.slo_document_name}"
+            )
+        slo_value = float(slo_value[1])
+        if isnan(slo_value):
+            raise InvalidSLOValue(
+                f"invalid format for SLO: {slo_name} of SLO document: {self.slo_document_name}"
+            )
+        return slo_value
+
+    def _get_slo_details(self, slo: SLODocumentSLOV1) -> SLODetails:
+        """
+        Build SLODetail object by retriving the current_value of the SLO from prometheus.
+        """
+        template = jinja2.Template(slo.expr)
+        prom_query = template.render({"window": slo.slo_parameters.window})
+        current_slo_response = self._get(
+            url=PROM_QUERY_URL, params={"query": (f"{prom_query}")}
+        )
+        slo_details = SLODetails(
+            namespace_name=self.namespace.namespace.name,
+            slo_document_name=self.slo_document_name,
+            cluster_name=self.namespace.namespace.cluster.name,
+            service_name=self.namespace.namespace.app.name,
+            slo=slo,
+            current_slo_value=self._extract_current_slo_value(
+                current_slo_response, slo.name
+            ),
+        )
+        return slo_details
+
+    def get_slo_details_list(self) -> list[SLODetails]:
+        try:
+            slo_details_list = threaded.run(
+                self._get_slo_details,
+                self.slos,
+                self.thread_pool_size,
+                return_exceptions=True,
+            )
+        finally:
+            self.cleanup()
```

Comment:
why need cleanup here? then this client can't send any more requests out, this should be done by caller